File: ClassEval_T/cpp/solution/MahjongConnect.cpp

#include <algorithm>

#include <vector>
#include <utility>
#include <set>
#include <cstdlib>

class MahjongConnect {
public:
    MahjongConnect(const std::pair<int, int>& boardSize, const std::vector<char>& icons);

    bool is_valid_move(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2);
    void remove_icons(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2);
    bool is_game_over() const;

    std::pair<int, int> BOARD_SIZE;
    std::vector<char> ICONS;
    std::vector<std::vector<char>> board;

    std::vector<std::vector<char>> create_board();
    bool has_path(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2) const;
};


MahjongConnect::MahjongConnect(const std::pair<int, int>& boardSize, const std::vector<char>& icons)
    : BOARD_SIZE(boardSize), ICONS(icons), board(create_board()) {}

std::vector<std::vector<char>> MahjongConnect::create_board() {
    std::vector<std::vector<char>> newBoard(BOARD_SIZE.first, std::vector<char>(BOARD_SIZE.second));

    for (int i = 0; i < BOARD_SIZE.first; ++i) {
        for (int j = 0; j < BOARD_SIZE.second; ++j) {
            newBoard[i][j] = ICONS[rand() % ICONS.size()];
        }
    }

    return newBoard;
}

bool MahjongConnect::is_valid_move(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2) {
    int x1 = pos1.first, y1 = pos1.second;
    int x2 = pos2.first, y2 = pos2.second;

    
    if (!(0 <= x1 && x1 < BOARD_SIZE.first && 0 <= y1 && y1 < BOARD_SIZE.second &&
        0 <= x2 && x2 < BOARD_SIZE.first && 0 <= y2 && y2 < BOARD_SIZE.second)) {
        return false;
    }

    
    if (pos1 == pos2) {
        return false;
    }

    
    if (board[x1][y1] != board[x2][y2]) {
        return false;
    }

    
    return has_path(pos1, pos2);
}

bool MahjongConnect::has_path(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2) const {
    std::set<std::pair<int, int>> visited;
    std::vector<std::pair<int, int>> stack;
    stack.push_back(pos1);

    while (!stack.empty()) {
        std::pair<int, int> currentPos = stack.back();
        stack.pop_back();

        if (currentPos == pos2) {
            return true;
        }

        if (visited.find(currentPos) != visited.end()) {
            continue;
        }

        visited.insert(currentPos);
        int x = currentPos.first, y = currentPos.second;

        
        for (const auto& direction : std::vector<std::pair<int, int>>{ {0, 1}, {0, -1}, {1, 0}, {-1, 0} }) {
            int newX = x + direction.first, newY = y + direction.second;
            if (0 <= newX && newX < BOARD_SIZE.first && 0 <= newY && newY < BOARD_SIZE.second) {
                if (visited.find({ newX, newY }) == visited.end() && board[newX][newY] == board[x][y]) {
                    stack.push_back({ newX, newY });
                }
            }
        }
    }

    return false;
}

void MahjongConnect::remove_icons(const std::pair<int, int>& pos1, const std::pair<int, int>& pos2) {
    board[pos1.first][pos1.second] = ' ';
    board[pos2.first][pos2.second] = ' ';
}

bool MahjongConnect::is_game_over() const {
    for (const auto& row : board) {
        if (std::any_of(row.begin(), row.end(), [](char icon) { return icon != ' '; })) {
            return false;
        }
    }
    return true;
}

File: ClassEval_T/java/solutuon/MahjongConnect.java
package org.example;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class MahjongConnect {
    int[] BOARD_SIZE;
    List<String> ICONS;
    String[][] board;

    public MahjongConnect(int[] BOARD_SIZE, List<String> ICONS) {
        this.BOARD_SIZE = BOARD_SIZE;
        this.ICONS = ICONS;
        this.board = createBoard();
    }

    public String[][] createBoard() {
        String[][] board = new String[BOARD_SIZE[0]][BOARD_SIZE[1]];
        Random random = new Random();
        for (int i = 0; i < BOARD_SIZE[0]; i++) {
            for (int j = 0; j < BOARD_SIZE[1]; j++) {
                board[i][j] = ICONS.get(random.nextInt(ICONS.size()));
            }
        }
        return board;
    }

    public boolean isValidMove(int[] pos1, int[] pos2) {
        int x1 = pos1[0], y1 = pos1[1];
        int x2 = pos2[0], y2 = pos2[1];

        if (!isValidPosition(x1, y1) || !isValidPosition(x2, y2)) {
            return false;
        }

        if (x1 == x2 && y1 == y2) {
            return false;
        }

        if (!board[x1][y1].equals(board[x2][y2])) {
            return false;
        }

        return hasPath(pos1, pos2);
    }

    private boolean isValidPosition(int x, int y) {
        return x >= 0 && x < BOARD_SIZE[0] && y >= 0 && y < BOARD_SIZE[1];
    }

    public boolean hasPath(int[] pos1, int[] pos2) {
        int x1 = pos1[0], y1 = pos1[1];
        int x2 = pos2[0], y2 = pos2[1];

        if (!isValidPosition(x1, y1) || !isValidPosition(x2, y2)) {
            return false;
        }

        boolean[][] visited = new boolean[BOARD_SIZE[0]][BOARD_SIZE[1]];
        return dfs(x1, y1, pos2, visited);
    }

    private boolean dfs(int x, int y, int[] target, boolean[][] visited) {
        if (x == target[0] && y == target[1]) {
            return true;
        }

        if (visited[x][y]) {
            return false;
        }

        visited[x][y] = true;

        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (int[] dir : directions) {
            int newX = x + dir[0];
            int newY = y + dir[1];
            if (isValidPosition(newX, newY) && !visited[newX][newY] && board[newX][newY].equals(board[x][y])) {
                if (dfs(newX, newY, target, visited)) {
                    return true;
                }
            }
        }

        return false;
    }

    public void removeIcons(int[] pos1, int[] pos2) {
        board[pos1[0]][pos1[1]] = " ";
        board[pos2[0]][pos2[1]] = " ";
    }

    public boolean isGameOver() {
        for (int i = 0; i < BOARD_SIZE[0]; i++) {
            for (int j = 0; j < BOARD_SIZE[1]; j++) {
                if (!board[i][j].equals(" ")) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        int[] boardSize = {4, 4};
        List<String> icons = new ArrayList<>();
        icons.add("a");
        icons.add("b");
        icons.add("c");

        MahjongConnect mc = new MahjongConnect(boardSize, icons);
        System.out.println(mc.isValidMove(new int[]{0, 0}, new int[]{1, 0}));
        mc.removeIcons(new int[]{0, 0}, new int[]{1, 0});
        System.out.println(mc.isGameOver());
    }
}